mila_datamodules/cli/prepare_torchvision.py

```diff
@@ -146,56 +256,286 @@ def __call__(
         return super()(root, *constructor_args, **constructor_kwargs)
 
 
+class Compose(AbstractPrepareVisionDataset[VD, P]):
+    def __init__(self, callables: list[AbstractPrepareVisionDataset]) -> None:
+        self.callables = callables
+
+    @runs_on_local_main_process_first
+    def __call__(
+        self,
+        root: str | Path = SLURM_TMPDIR / "datasets",
+        *dataset_args: P.args,
+        **dataset_kwargs: P.kwargs,
+    ) -> str:
+        for c in self.callables:
+            # TODO: Check that nesting `runs_on_local_main_process_first` decorators isn't a problem.
+            root = c(root, *dataset_args, **dataset_kwargs)
+        return root
+
+
 # NOTE: For some datasets, we have datasets stored in folders with the same structure. This here is
 # only really used to prevent repeating a bit of code in the dictionary below.
 # TODO: Find an exception to this rule and design this dict with that in mind.
-standardized_torchvision_dataset_dirs = {
+standardized_torchvision_datasets_dirs = {
     Cluster.Mila: "/network/datasets",
-    Cluster.Beluga: "/project/rpp-bengioy/data/curated",
+    Cluster.Beluga: "~/project/rpp-bengioy/data/curated",
 }
 
 prepare_torchvision_datasets: dict[type, dict[Cluster, PrepareVisionDataset]] = {
-    tvd.MNIST: {
-        # On the Mila and Beluga cluster we have archives which are extracted into 4 "raw" binary
-        # files. We do need to match the expected directory structure of the torchvision MNIST
-        # dataset though.
-        # NOTE: On Beluga, we also have the MNIST 'raw' files in
-        # /project/rpp-bengioy/data/MNIST/raw, no archives.
-        cluster: SymlinkArchives(
-            tvd.MNIST,
-            {
-                f"MNIST/raw/{archive.name}": archive
-                for archive in Path(f"{dataset_folder}/mnist").glob("*.gz")
-            },
-        )
-        for cluster, dataset_folder in standardized_torchvision_dataset_dirs.items()
+    tvd.Caltech101: {
+        cluster: Compose([
+            SymlinkDatasetFiles(
+                source=f"{datasets_folder}/caltech101",
+            ),
+            MoveFiles(
+                # Torchvision will look into a caltech101 directory to
+                # preprocess the dataset
+                files=[("*", "caltech101/*")]
+            ),
+            PrepareVisionDataset(tvd.Caltech101)
+        ])
```

Comment:
I really like this!
Perhaps we could rename `PrepareVisionDataset` to just `CallDatasetConstructor` or something like that? That way we could replace the `Abstract(...)` with `PrepareVisionDataset`.

Reply:
Agree